scilpy/tracking/propagator.py

```diff
@@ -163,47 +199,89 @@ def is_voxmm_in_bound(self, pos, origin):
 
         Parameters
         ----------
         pos: ndarray (3,)
             Current tracking position.
         v_in: ndarray (3,)
             Previous tracking direction.
+
+        Return
+        -------
+        direction: ndarray (3,)
+            A valid tracking direction. None if no valid direction is found.
+            Direction should be normalized.
         """
         raise NotImplementedError
 
 
-class ProbabilisticODFPropagator(AbstractPropagator):
-    """
-    Probabilistic direction tracker.
-
-    Parameters
-    ----------
-    tracking_field: scilpy tracking field object
-        The TrackingField object on which the tracking is done.
-    step_size: float
-        The step size for tracking.
-    rk_order: int
-        Order for the Runge Kutta integration.
-    """
-    def __init__(self, tracking_field: ODFField, step_size,
-                 rk_order):
-        super(ProbabilisticODFPropagator, self).__init__(
-            tracking_field, step_size, rk_order)
+class PropagatorOnSphere(AbstractPropagator):
+    def __init__(self, dataset, step_size, rk_order, dipy_sphere):
+        """
+        Parameters
+        ----------
+        dataset: scilpy.image.datasets.DataVolume
+            Trackable Dataset object.
+        step_size: float
+            The step size for tracking.
+        rk_order: int
+            Order for the Runge Kutta integration.
+        dipy_sphere: string, optional
+            If necessary, name of the DIPY sphere object to use to evaluate
+            directions.
+        """
+        super().__init__(dataset, step_size, rk_order)
+
+        if 'symmetric' not in dipy_sphere:
+            raise ValueError('Sphere must be symmetric. Call to '
+                             'get_opposite_direction will fail.')
+        self.sphere = dipy.data.get_sphere(dipy_sphere)
+        self.dirs = np.zeros(len(self.sphere.vertices), dtype=np.ndarray)
+        for i in range(len(self.sphere.vertices)):
+            self.dirs[i] = TrackingDirection(self.sphere.vertices[i], i)
+
+    def prepare_forward(self, seeding_pos):
+        raise NotImplementedError
+
+    def prepare_backward(self, line, forward_dir):
+        """
+        Called at the beginning of backward tracking, in case we need to
+        reset some parameters
+
+        Parameters
+        ----------
+        line: List
+            Result from the forward tracking, reversed.
+        forward_dir: ndarray (3,)
+            v_in chosen at the forward step.
+
+        Returns
+        -------
+        v_in: ndarray (3,)
+            Last direction of the streamline, of if it contains only the
+            seeding point (forward tracking failed), simply inverse the
+            forward direction.
+        """
+        if len(line) > 1:
+            last_dir = line[-1] - line[-2]
+            ind = self.sphere.find_closest(last_dir)
+            return TrackingDirection(last_dir, ind)
```

Comment:
Could it be possible that the closest direction returned by `sphere.find_closest` is not exactly `last_dir`? In what case the returned `TrackingDirection` would not be entirely accurate. Does this matter? Should we return `TrackingDirection(self.sphere.vertices[ind], ind)` just to be sure?

Reply:
Hmm. Yeah ok. I'll change it but leaving this conversation open in case anyone objects.